packages/ts-interface-generator/DEVELOPMENT.md

```diff
@@ -0,0 +1,55 @@
+- This AST is transformed to a nicely formatted string and written to a \*.generated.tsinterface.ts file next to the file in which the respective class was defined.
+
+### npm Publishing
+
+### Watch mode
+
+#### In Theory
+
+Main task: re-generate control interfaces once
+
+1. a file with metadata has been added
+2. metadata has been added to a file not covered so far
+3. metadata in an already covered file has changed
+4. metadata has been deleted from a file covered so far
+5. a file with metadata has been deleted
+
+Files to watch:
+
+1. new ts file creation
+2. content change in known ts file without metadata
+3. content change in known ts file with metadata
+4. content change in known ts file with metadata
+5. ts file deletion
+
+### In Practice
+
+The TypeScript compiler API offers a watch mode (or rather 2-3 flavors) which do not offer detailed change information, but are an apparently really optimized way of avoiding the multi-second TS initialization on every change. There are several callback hooks which need to be used in combination:
+
+1.  reportWatchStatusChanged is called with initially NO program, later an old program, and undefined error count and e.g. "message TS6032: File change detected. Starting incremental compilation..."
+2.  createProgram notifies that a change was detected
+3.  afterProgramCreate is called with the new Program (needed to create a TypeChecker!), but it is not yet known whether there are errors
+4.  reportDiagnostic is called for each error; the file name can be derived for each
+5.  reportWatchStatusChanged is then called with an error summary, the error count and e.g. "message TS6194: Found 4 errors. Watching for file changes." and can still access the new updated program
+
+There does not seem to be a way to get informed about deleted files.<br>
+New files and changed files can be enumerated by calling `program.getSemanticDiagnosticsOfNextAffectedFile()`. But this only works in the `afterProgramCreate` callback. By the time `reportWatchStatusChanged` is called, the "affected files" have already been handled and can no longer be accessed like this.
+
+The program object does have a `getState()` method and the state has a `changedFilesSet` property which contains the changed files, but those are not a public API, so they are not used.
+
+So the information about file changes is limited, but it is anyway questionable whether only changed files need to be considered when re-generating the interfaces: a change in a base class or in the UI5 type definitions could also influence the interface of an unchanged control implementation file. As performance for the overall standad flow is REALLY good (~20ms for a re-scan of all modules in UI5 and analyzing the types in the project files), there does not seem to be a need to restrict interface generation to those which MIGHT have an actual change.
```

Comment:
typo: 

`overall standad flow is REALLY good`

=> standard

`overall standard flow is REALLY good`

Reply:
done